functorch/csrc/BatchRulesBinaryOps.cpp

```diff
@@ -89,6 +89,48 @@ struct BinaryPointwiseBatchRuleHelper<F, Func, typelist<T1, T2, T...>> {
       &fn,\
       c10::guts::function_traits<decltype(fn)>::parameter_types>::apply)
 
+template <typename A, A a, typename C>
+struct BinaryRandomPointwiseBatchRuleHelper;
+
+template <typename F, F Func, typename T1, typename T2, typename... T>
+struct BinaryRandomPointwiseBatchRuleHelper<F, Func, typelist<T1, T2, T...>> {
+  static Tensor apply(const Tensor& tensor, const Tensor& other, T... extra_args) {
+    c10::impl::ExcludeDispatchKeyGuard guard(kVmapModeKey);
+    auto maybe_layer = maybeCurrentDynamicLayer();
+    auto cur_level = maybe_layer->layerId();
+    RandomnessType randomness = maybe_layer->randomness();
+
+    Tensor tensor_value;
+    optional<int64_t> tensor_bdim;
+    std::tie(tensor_value, tensor_bdim) = unwrapTensorAtLevel(tensor, cur_level);
+
+    Tensor other_value;
+    optional<int64_t> other_bdim;
+    std::tie(other_value, other_bdim) = unwrapTensorAtLevel(other, cur_level);
+
+    check_randomness(randomness, (tensor_bdim || other_bdim));
+    if (randomness == RandomnessType::Different && !tensor_bdim && !other_bdim) {
+      auto shape = tensor_value.sizes();
+      VmapDimVector shapeVec(shape.begin(), shape.end());
+      shapeVec.insert(shapeVec.begin(), maybe_layer->batchSize());
+      tensor_value = tensor_value.unsqueeze(0).expand(shapeVec);
+      tensor_bdim = 0;
```

Comment:
Do we even need this case? Does `auto res = _binary_pointwise_batch_rule<F, Func, T...>(` handle this?

It looks like what we're doing here is going to be faster than `auto res = _binary_pointwise_batch_rule<F, Func, T...>(` so it's probably good to keep this around, but just to check my understanding

Reply:
Just in case I misunderstood, putting both special cases here:
For the different, neither element is batched case: I believe that `_binary_pointwise_batch_rule` assumes that at least one of the tensors is batched (i.e. it never calls like `ensure_has_batch_dimension` or has a check that at least one bdim exists). So this basically just chooses the first tensor and expands that to make it batched. It probably could be done with `ensure_has_batch_dimension` though just to use those helper functions

For the same case: we could, we would just need some special handling again since `_binary_pointwise_batch_rule` always returns a bdim of 0 (it assumes the outcome must be batched). This does avoid all the checks + calls to helper functions (even if they won't do anything in this case)

- [ ] Add a couple comments explaining the reasoning for each case (speed, batch rule doesn't add a batch dim) @samdow